lib/client/client.go

```diff
@@ -548,58 +550,221 @@ func (client *NodeClient) scp(scpCommand scp.Command, shellCmd string, errWriter
 	return trace.Wrap(err)
 }
 
+func proxyConnection(client *NodeClient, incoming net.Conn, remoteAddr string) {
+	defer incoming.Close()
+	var (
+		conn net.Conn
+		err  error
+	)
+	log.Debugf("nodeClient.proxyConnection(%v -> %v) started", incoming.RemoteAddr(), remoteAddr)
+	for attempt := 1; attempt <= 5; attempt++ {
+		conn, err = client.Client.Dial("tcp", remoteAddr)
+		if err != nil {
+			log.Errorf("Connection attempt %v: %v", attempt, err)
+			// failed to establish an outbound connection? try again:
+			time.Sleep(time.Millisecond * time.Duration(100*attempt))
+			continue
+		}
+		// connection established: continue:
+		break
+	}
+	// permanent failure establishing connection
+	if err != nil {
+		log.Errorf("Failed to connect to node %v", remoteAddr)
+		return
+	}
+	defer conn.Close()
+	// start proxying:
+	doneC := make(chan interface{}, 2)
+	go func() {
+		io.Copy(incoming, conn)
+		doneC <- true
+	}()
+	go func() {
+		io.Copy(conn, incoming)
+		doneC <- true
+	}()
+	<-doneC
+	<-doneC
+	log.Debugf("nodeClient.proxyConnection(%v -> %v) exited", incoming.RemoteAddr(), remoteAddr)
+}
+
 // listenAndForward listens on a given socket and forwards all incoming connections
 // to the given remote address via
 func (client *NodeClient) listenAndForward(socket net.Listener, remoteAddr string) {
 	defer socket.Close()
 	defer client.Close()
-	proxyConnection := func(incoming net.Conn) {
-		defer incoming.Close()
-		var (
-			conn net.Conn
-			err  error
-		)
-		log.Debugf("nodeClient.listenAndForward(%v -> %v) started", incoming.RemoteAddr(), remoteAddr)
-		for attempt := 1; attempt <= 5; attempt++ {
-			conn, err = client.Client.Dial("tcp", remoteAddr)
-			if err != nil {
-				log.Errorf("Connection attempt %v: %v", attempt, err)
-				// failed to establish an outbound connection? try again:
-				time.Sleep(time.Millisecond * time.Duration(100*attempt))
-				continue
-			}
-			// connection established: continue:
+	// request processing loop: accept incoming requests to be connected to nodes
+	// and proxy them to 'remoteAddr'
+	for {
+		incoming, err := socket.Accept()
+		if err != nil {
+			log.Error(err)
 			break
 		}
-		// permanent failure establishing connection
+		go proxyConnection(client, incoming, remoteAddr)
+	}
+}
+
+func readByte(reader io.Reader) (byte, error) {
+	buf := []byte{0}
+	_, err := io.ReadFull(reader, buf)
+	return buf[0], err
+}
+
+const (
+	socks4Version                 byte = 0x04
+	socks5Version                 byte = 0x05
+	socks5Reserved                byte = 0x00
+	socks5AuthNotRequired         byte = 0x00
+	socks5AuthNoAcceptableMethods byte = 0xFF
+	socks5CommandConnect          byte = 0x01
+	socks5AddressTypeIPv4         byte = 0x01
+	socks5AddressTypeDomainName   byte = 0x03
+	socks5AddressTypeIPv6         byte = 0x04
+	socks5Succeeded               byte = 0x00
+)
+
+func socks5ProxyAuthenticate(incoming net.Conn) error {
```

Comment:
Can you move this function (really any parsing related logic) to `lib/utils/socks.go` then just call the needed parts here?

Reply:
Can you also add comments explaining what this function does?